planner/core/rule_partition_processor.go

```diff
@@ -1575,13 +1576,32 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
+	exprCollation, err := expression.CheckAndDeriveCollationFromExprs(sctx, opName, col.RetType.EvalType(), con, p.data[0])
```

Comment:
I think this part can be simplified to
```go
	_, collation := op.CharsetAndCollation()
	if collation != p.partCol.RetType.Collate && !(collate.IsBinCollation(collation) && opName == ast.EQ) {
		return 0, len(p.data), false
	}
```

only focus on the collation of expression and partition column, no matter what the coercibility is.

Reply:
Thank you for the suggestion, fixed it and it also improved the pruning when the collation was explicitly set to the same as the column :)